crates/lace-bin/src/typechecking/ctx.rs

```diff
@@ -0,0 +1,173 @@
+use crate::ast_metadata::{Metadata, MetadataId};
+use crate::name_resolution::ResolvedNames;
+use crate::parser::ast::Ident;
+use crate::parser::span::{Span, Spans};
+use crate::typechecking::ty::{PartialType, TypeVariable, TypeVariableGenerator};
+use bumpalo::Bump;
+use std::collections::HashMap;
+
+#[derive(Clone, Debug)]
+pub struct Types<'a> {
+    data: HashMap<TypeVariable, PartialType<'a>>,
+}
+
+impl<'a> Types<'a> {
+    pub fn new() -> Self {
+        Self {
+            data: Default::default(),
+        }
+    }
+
+    // pub fn add_type_variable(&mut self, var: TypeVariable) {
+    //     self.data
+    //         .insert(var, PartialType::Variable(var))
+    //         .unwrap_or_lice("duplicate type variable");
+    // }
+
+    pub fn type_of_type_variable(&self, var: TypeVariable) -> Option<&PartialType<'a>> {
+        // NOTE: the double lookup here is because of a limitation in Rust's borrow checking.
+        // NLL is not enough, and to make this a single get we need polonius. :(
+        match self.data.get(&var)? {
+            PartialType::Variable(v) if *v == var => self.data.get(&var),
+            PartialType::Variable(v) => self.type_of_type_variable(*v),
+            _ => self.data.get(&var),
+        }
```

Comment:
I don't really know what you're on about here? I can't even get an ownership error out of this:
```suggestion
        match self.data.get(&var)? {
            p @ PartialType::Variable(v) if *v == var => Some(p),
            PartialType::Variable(v) => self.type_of_type_variable(*v),
            p => Some(p),
        }
```
Or
```suggestion
        match self.data.get(&var) {
            p @ Some(PartialType::Variable(v)) if *v == var => p,
            Some(PartialType::Variable(v)) => self.type_of_type_variable(*v),
            p => p,
        }
```
Or
```suggestion
        match self.data.get(&var) {
            Some(PartialType::Variable(v)) if *v != var => self.type_of_type_variable(*v),
            p => p,
        }
```

Reply:
well, it was different, I think I changed something since which makes it possible.